cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -227,23 +229,39 @@ private predicate storeStepChi(Node node1, Content f, PostUpdateNode node2) {
       c = chi.getResultType() and
       exists(int startBit, int endBit |
         chi.getUpdatedInterval(startBit, endBit) and
-        f.(FieldContent).hasOffset(c, startBit, endBit)
+        f.hasOffset(c, startBit, endBit)
       )
       or
-      getWrittenField(store, f.(FieldContent).getAField(), c) and
-      f.(FieldContent).hasOffset(c, _, _)
+      getWrittenField(store, f.getAField(), c) and
+      f.hasOffset(c, _, _)
     )
   )
 }
 
+private predicate arrayStoreStepChi(Node node1, ArrayContent a, PostUpdateNode node2) {
+  a = TArrayContent() and
+  exists(StoreInstruction store |
+    node1.asInstruction() = store and
+    (
+      // `x[i] = taint()`
+      store.getDestinationAddress() instanceof PointerAddInstruction
+      or
+      // `*p = taint()`
+      store.getDestinationAddress().(CopyValueInstruction).getUnary() instanceof LoadInstruction
+    ) and
+    node2.asInstruction().(ChiInstruction).getPartial() = store
+  )
```

Comment:
Check `isResultConflated` here?

Reply:
It will always be a non-conflated `Chi` because the char preds for `ArrayStoreNode` and `PointerStoreNode` require the instruction to be non-conflated. Performance on `linux` seems to be the same, but if you think it's important for readability I can add the check. I've added a comment saying that it'll always be non-conflated in 50ad4cfec4e80387fa67e8f9a21eeb45128a20fd.

Reply:
I'm okay with the readability as long as the performance is the same.